pkg/ingester/tailer.go

```diff
@@ -0,0 +1,231 @@
+package ingester
+
+import (
+	"encoding/binary"
+	"fmt"
+	"hash/fnv"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/parser"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// This is to limit size of dropped stream to make it easier for querier to query.
+// While dropping stream we divide it into batches and use start and end time of each batch to build dropped stream metadata
+const maxDroppedStreamSize = 10000
+
+type tailer struct {
+	id       uint32
+	orgID    string
+	matchers []*labels.Matcher
+	regexp   *regexp.Regexp
+
+	sendChan chan *logproto.Stream
+	closed   bool
+
+	blockedAt      *time.Time
+	blockedMtx     sync.RWMutex
+	droppedStreams []*logproto.DroppedStream
+
+	conn logproto.Querier_TailServer
+}
+
+func newTailer(orgID, query, regex string, conn logproto.Querier_TailServer) (*tailer, error) {
+	matchers, err := parser.Matchers(query)
+	if err != nil {
+		return nil, err
+	}
+
+	var re *regexp.Regexp
+	if regex != "" {
+		re, err = regexp.Compile(regex)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &tailer{
+		orgID:          orgID,
+		matchers:       matchers,
+		regexp:         re,
+		sendChan:       make(chan *logproto.Stream, 2),
+		conn:           conn,
+		droppedStreams: []*logproto.DroppedStream{},
+		id:             generateUniqueID(orgID, query, regex),
```

Comment:
Hmm, does this mean one org/instance cannot tail for the same thing on two different laptops?

Reply:
I am adding current timestamp in ns and query details which makes chances of someone running same query at exact same time very very less likely. Do you want me to add some random bytes or something to it?
https://github.com/grafana/loki/blob/08dd6685501a83d3adc7a54f05dc71c2c4473021/pkg/ingester/tailer.go#L220-L231